packages/common/src/converter/types.ts

```diff
@@ -1,7 +1,36 @@
-import type * as iface from '@temporalio/proto/lib/coresdk';
-import { TextEncoder, TextDecoder } from './encoding';
+import type { coresdk } from '@temporalio/proto/lib/coresdk';
+import { TextDecoder, TextEncoder } from './encoding';
 
-export type Payload = iface.coresdk.common.IPayload;
+export type Payload = coresdk.common.IPayload;
+
+export interface EncodedPayload extends Payload {
+  encoded: true;
+}
+
+export interface DecodedPayload extends Payload {
+  decoded: true;
+}
+
+/** An object T with any nested values of type ToReplace replaced with ReplaceWith */
+export type ReplaceNested<T, ToReplace, ReplaceWith> = T extends (...args: any[]) => any
+  ? T
+  : [keyof T] extends [never]
+  ? T
+  : T extends { [k: string]: coresdk.common.IPayload } | null
```

Comment:
Do we need `| null`?

If we do then I'd do this:
```ts
  : T extends { [k: string]: ToReplace } | null
```

Otherwise I'd just replace with:
```ts
  : T extends { [k: string]: ToReplace }
```